cmd/runtimetest/main.go

```diff
@@ -320,6 +322,95 @@ func validateROPaths(spec *rspec.Spec) error {
 	return nil
 }
 
+func getIDMappings(idType string) ([]rspec.IDMapping, error) {
+	var mapFile string
+	var idMaps []rspec.IDMapping
+	switch idType {
+	case "uid_map":
+		mapFile = "/proc/1/uid_map"
+	case "gid_map":
+		mapFile = "/proc/1/gid_map"
+	}
+	f, err := os.Open(mapFile)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	s := bufio.NewScanner(f)
+	for s.Scan() {
+		if err := s.Err(); err != nil {
+			return nil, err
+		}
+
+		text := strings.Replace(strings.TrimSpace(s.Text()), "          ", " ", -1)
+		idMap := strings.Split(text, " ")
+		if len(idMap) == 3 {
+			hostID, err := strconv.ParseUint(idMap[0], 0, 32)
+			if err != nil {
+				return nil, err
+			}
+			containerID, err := strconv.ParseUint(idMap[1], 0, 32)
+			if err != nil {
+				return nil, err
+			}
+			mapSize, err := strconv.ParseUint(idMap[2], 0, 32)
+			if err != nil {
+				return nil, err
+			}
+			idMaps = append(idMaps, rspec.IDMapping{HostID: uint32(hostID), ContainerID: uint32(containerID), Size: uint32(mapSize)})
+		} else {
+			return nil, fmt.Errorf("invalid format in %v", mapFile)
+		}
+	}
+
+	return idMaps, nil
+}
+
+func validateUIDMappings(spec *rspec.Spec) error {
+	logrus.Debugf("validating uidMappings")
+	idMaps, err := getIDMappings("uid_map")
+	if err != nil {
+		return fmt.Errorf("can not get items from uid_map: %v", err)
+	}
+	for _, v := range spec.Linux.UIDMappings {
```

Comment:
On Wed, Sep 14, 2016 at 12:34:47AM -0700, Ma Shimiao wrote:
> As you said above, if user could setup the mappings on their own
> before calling ‘start’, I think we can't check this. But in my
> opinion, user can't setup such thing before calling start.

I think we want to make runtimetest too strict (with a caveat
explaining that [1]) and then not do the legal things it doesn't
allow.  The constraints on the validated runtime are much weaker if
you make runtimetest too permissive.  So where hitting “just right”
isn't possible, I'd err on the side of “too strict”.

[1]: https://github.com/opencontainers/ocitools/pull/211#discussion_r78610440